src/rust/ide/view/graph-editor/src/component/node/output/port.rs

```diff
@@ -0,0 +1,458 @@
+        set_port_count    (this,t:usize) { this.shape.port_count.set(t as f32) }
+        set_padding_left  (this,t:f32)   { this.shape.padding_left.set(t) }
+        set_padding_right (this,t:f32)   { this.shape.padding_right.set(t) }
+    }
+
+    fn events(&self) -> &component::ShapeViewEvents {
+        match self {
+            Self::Single (t) => &t.events,
+            Self::Multi  (t) => &t.events,
+        }
+    }
+}
+
+impl display::Object for PortShapeView {
+    fn display_object(&self) -> &display::object::Instance {
+        match self {
+            Self::Single (view) => view.display_object(),
+            Self::Multi  (view) => view.display_object(),
+        }
+    }
+}
+
+
+
+// =================
+// === Port Frp  ===
+// =================
+
+ensogl::define_endpoints! {
+    Input {
+        set_size_multiplier (f32),
+        set_definition_type (Option<Type>),
+        set_usage_type      (Option<Type>),
+    }
+
+    Output {
+        tp       (Option<Type>),
+        on_hover (bool),
+        on_press (),
+    }
+}
+
+#[derive(Clone,Debug,Default)]
+pub struct Model {
+    pub frp    : Option<Frp>,
+    pub shape  : Option<PortShapeView>,
+    pub index  : usize,
+    pub length : usize,
+}
+
+impl Model {
+    pub fn init_shape
+    (&mut self, logger:impl AnyLogger, scene:&Scene, styles:&StyleWatch, port_index:usize, port_count:usize)
+    -> (PortShapeView,Frp) {
+        let logger_name = format!("port({},{})",self.index,self.length);
+        let logger      = Logger::sub(logger,logger_name);
+        let shape       = PortShapeView::new(port_count,&logger,scene);
+
+        let is_first      = port_index == 0;
+        let is_last       = port_index == port_count.saturating_sub(1);
+        let padding_left  = if is_first { -INFINITE } else {  SEGMENT_GAP_WIDTH / 2.0 };
+        let padding_right = if is_last  {  INFINITE } else { -SEGMENT_GAP_WIDTH / 2.0 };
+        shape.set_index(port_index);
+        shape.set_port_count(port_count);
+        shape.set_padding_left(padding_left);
+        shape.set_padding_right(padding_right);
+
+        self.init_frp(&shape,styles);
+        self.shape = Some(shape);
+        (self.shape.as_ref().unwrap().clone_ref(),self.frp.as_ref().unwrap().clone_ref())
+    }
+
+    fn init_frp(&mut self, shape:&PortShapeView, styles:&StyleWatch) {
+        let frp     = Frp::new();
+        let network = &frp.network;
+        let events  = shape.events();
+        let opacity = Animation::<f32>::new(network);
+        let color   = color::Animation::new(network);
+
+        frp::extend! { network
+
+            // === Mouse Event Handling ===
+
+            frp.source.on_hover <+ bool(&events.mouse_out,&events.mouse_over);
+            frp.source.on_press <+ events.mouse_down;
+
+
+            // === Opacity ===
+
+            opacity.target <+ events.mouse_over.constant(PORT_OPACITY_HOVERED);
+            opacity.target <+ events.mouse_out.constant(PORT_OPACITY_NOT_HOVERED);
```

Comment:
Should the values `PORT_OPACITY_HOVERED` and `PORT_OPACITY_NOT_HOVERED` be defined in the theme?

Reply:
That's a good idea! I'll add coment there and I'll fix it during theme re-rework today , as themes got broken recently

Reply:
https://github.com/enso-org/ide/issues/1044